sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    internal class OperationInternal<TResult>
+    {
+        private readonly TimeSpan _defaultPollingInterval;
+
+        private readonly IOperation<TResult> _operation;
+
+        private TResult _value;
+
+        public OperationInternal(string operationTypeName,
+            ClientDiagnostics clientDiagnostics,
+            IOperation<TResult> operation,
+            TimeSpan? defaultPollingInterval = null) : this(operationTypeName, clientDiagnostics, defaultPollingInterval)
+        {
+            _operation = operation;
+        }
+
+        protected OperationInternal(string operationTypeName,
+            ClientDiagnostics clientDiagnostics,
+            TimeSpan? defaultPollingInterval)
+        {
+            OperationTypeName = operationTypeName;
+            ClientDiagnostics = clientDiagnostics;
+            _defaultPollingInterval = defaultPollingInterval ?? TimeSpan.FromSeconds(1);
+        }
+
+        public bool HasValue { get; private set; }
+
+        public bool HasCompleted { get; set; }
+
+        public TResult Value
+        {
+            get
+            {
+                if (HasValue)
+                {
+                    return _value;
+                }
+                else if (OperationFailedException != null)
+                {
+                    throw OperationFailedException;
+                }
+                else
+                {
+                    throw new InvalidOperationException("The operation has not completed yet.");
+                }
+            }
+            set
+            {
+                _value = value;
+                HasValue = true;
+            }
```

Comment:
I'll probably add a check here to throw if we try to set the value to null by accident.
We're making `Value` settable because we have operations that change `Value` in every status update.